camkes/templates/component.common.c

```diff
@@ -42,6 +42,18 @@
 
 /*? macros.show_includes(me.type.includes) ?*/
 
+/*- if 'affinity' in configuration[me.name].keys() -*/
+#if(/*? configuration[me.name].get('affinity') ?*/ >= CONFIG_MAX_NUM_NODES)
+#error "Invalid Affinity Value of /*? configuration[me.name].get('affinity') ?*/"
+#endif
+/*- endif -*/
+
+/*- if 'num_vcpus' in configuration[me.name].keys() -*/
+#if(/*? configuration[me.name].get('num_vcpus') ?*/ > CONFIG_MAX_NUM_NODES)
+#error "Invalid Number of VCPUs"
```

Comment:
Do we really want to restrict this? Having more vCPUs than physical cores might not be optimal, but it is still a legal configuration and there might be use cases to run a legacy VM.

Reply:
This restricts the number of vCPUs per VM so there can still be more vCPUs than physical cores in a whole system just not within an individual VM. So for example in a 4 core platform, VM0 could have 4 vCPUs and VM1 could have 2 vCPUs totaling to 6 vCPUs in the system but neither VM has more vCPUs than the number of physical cores. If what is running inside a VM was designed to run on the platform natively, I wouldn't expect it to want or need more vCPUs than the number of physical cores the platform provides.

Reply:
But that was my point, there might be VMs what want more vCPUs than physical cores for whatever reason (might also just be for testing). I would also not take the assumption that the VM is always a VM build exactly for that platform, it might be a generic VM also. Using more vCPUs is is not optimal, but it's also not an illegal. So we should not block this. We could print a warning that this is not an optimal configuration.

Reply:
Ah I see. In that case additional changes would need to be made to the VM_Arm component see
https://github.com/seL4/camkes-vm/blob/master/components/VM_Arm/src/main.c#L1229
This change currently just extends that runtime check to be a compile time check. I am not opposed to changing it to a warning but maybe it makes sense to keep it as an error until that runtime functionality changes. Or maybe it should be a warning that states the num_vcpus will be capped to MAX_NUM_NODES. Thoughts?

Reply:
Ok, you have a point there. Then this is currently illegal within the implementation, indeed, and thus an error - until we extend things to support this.